core/server/services/auth/api-key/admin.js

```diff
@@ -0,0 +1,97 @@
+const jwt = require('jsonwebtoken');
+const models = require('../../../models');
+const {
+    InternalServerError,
+    UnauthorizedError
+} = require('../../../lib/common/errors');
+
+const JWT_OPTIONS = {
+    maxAge: '5m',
+    algorithms: ['HS256']
+};
+
+/**
+ * Remove 'Bearer' from raw authorization header and extract the JWT token.
+ * Eg. Authorization: Bearer ${JWT}
+ * @param {string} header
+ */
+const extractTokenFromHeader = function extractTokenFromHeader(header) {
+    const [scheme, token] = header.split(' ');
+
+    if (/^Bearer$/i.test(scheme)) {
+        return token;
+    }
+
+    return;
+};
+
+const authenticateAdminAPIKey = function authenticateAdminAPIKey(req, res, next) {
+    // allow fallthrough to other auth methods or final ensureAuthenticated check
+    if (!req.headers || !req.headers.authorization) {
+        return next();
+    }
+
+    if (req.query && req.query.content_key) {
+        return next(new UnauthorizedError({
+            message: 'Admin API does not support query param authentication',
+            code: 'INVALID_AUTH_TYPE'
+        }));
+    }
+
+    const token = extractTokenFromHeader(req.headers.authorization);
+
+    if (!token) {
+        return next(new UnauthorizedError({
+            message: 'Authorization header format is "Authorization: Bearer [token]"',
+            code: 'INVALID_AUTH_HEADER'
+        }));
+    }
+
+    const decoded = jwt.decode(token, {complete: true});
+    const apiKeyId = decoded.header.kid;
+
+    models.ApiKey.findOne({id: apiKeyId}).then((apiKey) => {
+        if (!apiKey) {
+            return next(new UnauthorizedError({
+                message: 'Unknown Admin API Key',
+                code: 'UNKNOWN_ADMIN_API_KEY'
+            }));
+        }
+
+        if (apiKey.get('type') !== 'admin') {
+            return next(new UnauthorizedError({
+                message: 'Incorrect API Key type',
+                code: 'INCORRECT_API_KEY_TYPE'
+            }));
+        }
+
+        const secret = Buffer.from(apiKey.get('secret'), 'hex');
+        const options = Object.assign({
+            // ensure the token was meant for this endpoint
+            aud: req.originalUrl
```

Comment:
@allouis do you think this is enough or should we also check the host? By this stage we're already failing if the key is unknown so I'm wondering if additional checks here will be unnecessary barriers that could cause issues with odd proxy setups

Reply:
I think this is fine for now :+1: We can iron out any other stuff later